source/controllers/tests/testDissipativeImpedanceController.cpp

```diff
@@ -0,0 +1,106 @@
+#include "controllers/impedance/Dissipative.hpp"
+#include <gtest/gtest.h>
+
+using namespace controllers::impedance;
+using namespace StateRepresentation;
+
+class DissipativeImpedanceControllerTest : public testing::Test {
+protected:
+  void SetUp() override {}
+
+  void set_controller_space(const ComputationalSpaceType& computational_space) {
+    controller_ = Dissipative(computational_space);
+  }
+
+  Dissipative controller_;
+};
+
+TEST_F(DissipativeImpedanceControllerTest, TestOrthonormalize) {
+  Eigen::Matrix3d basis = Eigen::Matrix3d::Random();
+  Eigen::Vector3d eigenvector = Eigen::Vector3d::Random();
+
+  // compute the orthonormal basis
+  Eigen::Matrix3d orthonormal_basis = controller_.compute_orthonormal_basis(basis, eigenvector);
+
+  // first column should the normalized eigenvector
+  Eigen::Vector3d err = orthonormal_basis.col(0) - eigenvector.normalized();
+  for (int i = 0; i < 3; ++i) { EXPECT_NEAR(err(i), 0., 10e-4); }
+  // all inner products must be equal to 0
+  EXPECT_NEAR(orthonormal_basis.col(0).dot(orthonormal_basis.col(1)), 0., 10e-4);
+  EXPECT_NEAR(orthonormal_basis.col(1).dot(orthonormal_basis.col(2)), 0., 10e-4);
+  EXPECT_NEAR(orthonormal_basis.col(0).dot(orthonormal_basis.col(2)), 0., 10e-4);
+  // all the magnitude should be equal to 1
+  EXPECT_NEAR(orthonormal_basis.col(0).norm(), 1., 10e-4);
+  EXPECT_NEAR(orthonormal_basis.col(1).norm(), 1., 10e-4);
+  EXPECT_NEAR(orthonormal_basis.col(2).norm(), 1., 10e-4);
+}
+
+TEST_F(DissipativeImpedanceControllerTest, TestComputeDamping) {
+  Eigen::MatrixXd damping;
+  Eigen::Matrix3d identity3 = Eigen::Matrix3d::Identity();
+  Eigen::MatrixXd identity6 = Eigen::MatrixXd::Identity(6, 6);
+  Eigen::VectorXd eigenvector = Eigen::VectorXd::Random(6);
+  // case LINEAR
+  set_controller_space(ComputationalSpaceType::LINEAR);
+  controller_.compute_damping(eigenvector);
+  damping = controller_.get_damping();
+
+  // linear part is identity
+  for (int i = 0; i < 3; ++i) {
```

Comment:
Rather than indexed for loops to check for zero values, you could use either of the follow approaches:

```c++
#include <numeric>
auto block = damping.block<3,3>(0, 0);
EXPECT_NEAR(std::accumulate(block.data(), block.data()+block.size(), 0.0), 0.0, tolerance)
```

```c++
for(auto elem : damping.block<3,3>(0, 0).reshaped()) {
  EXPECT_NEAR(elem, block.data()+block.size(), 0.0), 0.0, tolerance)
}
```

Reply:
neat thanks